Add beta WP Story block

Tug opened the pull request:
#### Changes proposed in this Pull Request:
This PR introduces a new (Beta) story block for WordPress.com

The idea is that content creation will happen on the Apps, then we need a way to display stories and edit them from WordPress and we figured this could be done via a block living in jetpack.

A few screenshots:

| Story creation | Story rendered as a post |
|---------------|--------------------------|
|<img width="742" alt="Screenshot 2020-07-10 at 11 46 29" src="https://user-images.githubusercontent.com/230230/87141680-b461fd80-c2a3-11ea-8c33-b740b9bfd076.png">|<img width="787" alt="Screenshot 2020-07-10 at 11 47 38" src="https://user-images.githubusercontent.com/230230/87141687-b62bc100-c2a3-11ea-9ce9-202dc3f72d29.png">|

| Story playing  | Story played                      |
|---------------|--------------------------|
|<img width="1287" alt="Screenshot 2020-07-10 at 11 47 59" src="https://user-images.githubusercontent.com/230230/87141695-b7f58480-c2a3-11ea-9976-a52e1a927e1f.png">|<img width="782" alt="Screenshot 2020-07-10 at 11 48 16" src="https://user-images.githubusercontent.com/230230/87141699-ba57de80-c2a3-11ea-9d16-52d4a37c9c1a.png">|

#### Jetpack product discussion
pbD5rk-h
pbD5rk-dS-p2
p1HpG7-9Cp-p2

#### Does this pull request change what data or activity we track or use?
No

#### Review Instructions:

Review as a new block in `extensions/blocks/story`.
The block part and the player part should be reviewed separately as they are quite independent.

The block `edit` function simply mounts the player with some custom configuration (to not play the story in fullscreen for instance).
The block `save` function simply saves the block as an emply div where the player will be mounted. It also includes the list of media files in the block attributes.

Thus rendering the block on the client is fully dynamic. It allows us to evolve Stories while maintaining backward compatibility. In the future we might want to host stories on wordpress.com so they can easily be shared as embeds. In that scenario, the rendering will simply be an iframe.

#### Testing instructions:
- Try creating a story from the block editor (this is not the main flow the user will go through so this is not super polished at the moment). Type "Story", click edit and select Media, try to include a video as well
- View the created story on your site, try to play, pause, mute, unmute, navigate through the slides, replay, exit fullscreen, resize your screen. Try on mobile
- Notice the story renders in full screen on load when displayed as a post

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Introduce new experimental "WP Story" block

#### TODO
Here is a list of improvements that might or might not make it for the beta:
- [ ] Handle multiple image ratios: crop top and bottom of the images/videos to avoid showing black margins on the side. If more than X% of the media is hidden start displaying margins on the top and bottom
- [ ] Fix scrolling to the top after story playback
- [ ] Fix video not playing on some mobile browsers or is choppy
- [ ] Other fixes for issues discovered during the pre-alpha (see Issue column https://github.com/orgs/Automattic/projects/191)

jeherve commented:
Do you think you could loop in Jetpack Product on this, as per this:
PCYsg-iiu-p2/#2-loop-in-the-jetpack-division

That should help us understand the story behind that block 😉

Tug commented:
@jeherve Thanks for the link. Yes the plan is to sync with the Jetpack Product team as soon as possible now that we are approaching "alpha". We have our own internal P2 on the project but granted this probably deserves it's own MT on Jetpack P2 👍 . Not sure at this point who should be involved, we haven't determined someone to review the work at this time.

Tug commented:
Btw, this should probably have been created as a draft PR, because although the code is mostly ready for reviewed I want to add details and screenshots to the PR description.

jeherve commented:
> Not sure at this point who should be involved, we haven't determined someone to review the work at this time.

Once this has been discussed with the Product folks and once the code is ready for review, I'd recommend having one person on your team looking at it and testing it. Once you're both happy with it, you can change the PR status to "Needs Review" and we in Crew will be happy to review it and help you get it shipped to Jetpack and WordPress.com Simple sites. 👍

Tug commented on package.json:
```diff
@@ -137,6 +137,7 @@
 		"gulp-sass": "4.1.0",
 		"gulp-sourcemaps": "2.6.5",
 		"gulp-tap": "2.0.0",
+		"htm": "3.0.4",
```
I just noticed that we're already using preact on jetpack, I had no idea!!

Maybe switching to JSX for the user facing part of the block makes more sense now... `htm` has one advantage though, I can be exposed to users so they can customize the views on their end. For instance [here](https://github.com/Automattic/jetpack/pull/16302/files#diff-a1ecaecd7e92d91bdb9f51c729281dc7) I defined the default views that can be customized. Sounds like a YAGNI though now that we have preact with JSX built in 🤔

jeherve replied:
We don't use Preact in the extensions, though; it's limited to a small part of the codebase (the Instant Search interface), so you wouldn't be able to rely on Preact for your block I'm afraid.

Tug replied:
Oh I see! The part that uses preact here isn't really part of the block actually and could be moved out of the extension if we want to. It's just the visitor-facing UI that renders the story on the site. Preact I'm pretty sure of as it fits our requirements for a lightweight UI framework.

jeherve replied:
Thanks for the extra details, that helps!

It does sound like Preact could indeed do the job. However, I am not sure it would be the best choice on the long term here. We've tried to stay as close as possible to Core with blocks so far:
- it allows us to benefit from the work the community does every time a component is created and updated.
- it integrates well with existing components and processes such as translations.
- it helps us making sure that whatever component updates will happen in the future can be adapted to Jetpack components.
- it makes it easier to onboard folks creating new blocks or picking up maintenance of a block.
- it allows us to benefit from existing building tools as well.

Moving away from those things isn't something I would personally recommend, as it can impact how the block may be maintained and involve in the future.

To add a bit of background, we picked Preact for the Instant Search interface because that's something that needs to be loaded on each and every page of a site. In this case, the size of the bundle we serve is extremely important, and after some tests we concluded that using Preact over WordPress components was worth it, even if we had to go through some loops to implement things such as i18n.

simison replied:
Including Preact and related libraries in bundle here _and_ in Search doubles up KBs it uses, too. ;-)

simison commented on extensions/blocks/story/edit.js:
```diff
@@ -0,0 +1,215 @@
+		const imageSizes = select( 'core/editor' ).getEditorSettings().imageSizes;
+		const resizedMediaFiles = props.attributes.ids.reduce( ( currentResizedMediaFiles, id ) => {
+			const image = select( 'core' ).getMedia( id );
+			const sizes = get( image, [ 'media_details', 'sizes' ] );
```
You could replace `get` with `image?.media_details?.sizes` if you're not sure all those steps exist.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

Tug replied:
Indeed that looks better 👍 

Btw, this file is heavily inspired by `slideshow/edit.js` and I'm still not sure what will keep, so a lot of code might go actually

simison commented on extensions/blocks/story/player/components/icons.scss:
```diff
@@ -0,0 +1,121 @@
+	url("data:application/font-woff2;base64,AAEAAAAOAIAAAwBgR0RFRgASAAoAAAZUAAAAFkdQT1MAGQAMAAAGbAAAABBHU1VCbIx0hQAABnwAAAAaT1MvMup2AjEAAAN4AAAAYGNtYXCk0d8LAAAD2AAAANhnYXNw//8AAwAABkwAAAAIZ2x5Zv93xxIAAADsAAAB4GhlYWQO2ulTAAADBAAAADZoaGVhBAECAgAAA1QAAAAkaG10eAQrAjMAAAM8AAAAFmxvY2EClAH3AAAC7AAAABZtYXhwAGQA4QAAAswAAAAgbmFtZRwPNeQAAASwAAABenBvc3T/hgAyAAAGLAAAACAAAgCAAGsBgAGVAAMABwAAATMRIyMRMxEBK1VVq1UBlf ...
+}
+
+.material-icons {
```
This seems little generic classname and could do with `jetpack_` prefix? 🤔

simison commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,129 @@
+		esc_attr( get_site_icon_url( 32, includes_url( 'images/w-logo-blue.png' ) ) ),
+		esc_html( get_bloginfo( 'name' ) ),
+		esc_html( get_bloginfo( 'description' ) ),
+		join( "\n", array_map( $get_slide_template, $attributes['mediaFiles'], array_keys( $attributes['mediaFiles'] ) ) )
```
Does this need escaping? Remember that attributes could contain anything really.

simison commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,129 @@
+							%s
+						</div>
+					</div>
+					<button class="wp-story-exit-fullscreen mdc-icon-button" aria-label="Exit Fullscreen" aria-pressed="false">
```
"Exit Fullscreen" needs `__()`

simison commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,129 @@
+
+	return sprintf(
+		'<div class="wp-block-jetpack-story wp-story aligncenter" data-settings="%s">
+			<div class="wp-block-jetpack-story_container wp-story-container" style="display: block; opacity: 1;">
```
Did you consider moving CSS to a file instead of inlining it?

Tug replied:
This part should be removed actually, thanks for pointing it out 👍

simison commented on extensions/blocks/story/edit.js:
```diff
@@ -0,0 +1,215 @@
+import Story from './story';
+import './editor.scss';
+
+const ALLOWED_MEDIA_TYPES = [ 'image', 'video' ];
```
What happens for video files if VideoPress is active?

Tug replied:
I haven't tested with videopress, but that's something we should do indeed 👍

simison replied:
Yup, that would be extremely important for .com sites, especially on Atomic.

simison commented on extensions/blocks/story/player/components/icons.scss:
```diff
@@ -0,0 +1,121 @@
+@font-face {
+	font-family: 'Material Icons for WP Stories';
```
I have zero context here but Have you considered simply including SVGs instead of a font?

SVG would be easier to maintain and better for web performance since it can be bundled and even shipped incrementally, vs everything at once.

Tug replied:
I did consider it but I'm not using react/JSX here so I didn't see an easy way to inline it into the code without changing the build.

A bit of context: anything that lives in `story/player/` right now could (and probably will) be moved to an external service later on as we'll need embeds to support sharing stories (Meaning the story will likely be displayed in an iframe loaded from wordpress.com). So it should be seen as a standalone app in itself whose only dependencies are `preact` and `htm`. I'm leveraging the jetpack build right now but I would like to avoid doing any changes to it.

About JSX vs HTM I'm considering converting the whole app to JSX. I'm a bit hesitant as `htm` can allow users to customize the UI without a building step which is quite nice but which is not a priority for us atm

simison replied:
Thanks for the context! It's helpful. :-)

One way to inline svg's by having them in JS files, like so:

https://github.com/WordPress/gutenberg/blob/6e80f44539a03d658935303000aecf313dba9127/packages/icons/src/library/align-left.js

Alternatively this works:

```jsx
import Image from './image.svg';

<img src={ Image } />
```

Tug replied:
I see but your first approach uses JSX.

The second one could work with htm but it will depend on the packager to either inline the image using base64 or move the assets to the right folder and generate the correct url in place. Since it seemed to me that the jetpack packager is doing the second option I did not go that route.

simison replied:
The build tooling could be changed so that you can have inlined svg. You've done quite a bit of work in getting things work in font (which I think isn't favorable for maintainability, flexibility and performance) and same effort could've instead one to build-tooling. Jetpack uses [`calypso-build`](https://github.com/Automattic/wp-calypso/tree/master/packages/calypso-build) but things can also be implemented directly at Jetpack since we extend the webpack config from that tool. Doing changes higher up in the source the better, of course.

jeherve replied:
Maybe we could do something like what we do for the Podcast player block instead?
https://github.com/Automattic/jetpack/blob/06e20fbb4eb098c32bc01631f570735b1f55e5ac/extensions/blocks/podcast-player/style.scss#L431-L438

For more info on this: #15857

simison commented on extensions/blocks/story/player/style.scss:
```diff
@@ -0,0 +1,377 @@
+	list-style: none;
+	padding: 0;
+	z-index: 1;
+	background-color: #0e1112;
```
Would be great to move HEXes to SASS variables somewhere for easier maintainability.

simison commented on extensions/blocks/story/player/style.scss:
```diff
@@ -0,0 +1,377 @@
+	&.player-container-rtl .wp-block-jetpack-story_button-prev,
+	.player-button-next.player-button-white,
+	.wp-block-jetpack-story_button-next {
+		background-image: url( "data:image/svg+xml,%3Csvg%20xmlns='http://www.w3.org/2000/svg'%20width='24'%20height='24'%20viewBox='0%200%2024%2024'%3E%3Cpath%20d='M5.88%204.12L13.76%2012l-7.88%207.88L8%2022l10-10L8%202z'%20fill='white'/%3E%3Cpath%20fill='none'%20d='M0 0h24v24H0z'/%3E%3C/svg%3E" );
```
Hm, don't we have any way to inline SVGs in build-pipeline in Sass? 🤔 Would be much easier to maintain this if it would be in separate file.

Tug replied:
This code is outdated as well, sorry I need to clean up. All the icons I use for the player are actually encoded into a font that's inlined into the css here https://github.com/Automattic/jetpack/pull/16302/files/c3dd77328091b2e3d6da6bd920a01c6fa00686cc#diff-4c86797267f3a6de9c98fb77dd3ddc95R10

simison replied:
Yup, I noticed the inspiration. :-) The slideshow was written some time ago (it also uses Classes vs hooks in functional component) so use your judgement.

Tiled gallery is another one where Slider originally took inspiration from, and Tiled gallery was largely copy of core gallery block. So I'd prolly check core gallery to make sure you have the greatest and latest. :-D

jeherve replied:
> it also uses Classes vs hooks in functional component) so use your judgement.

Since you mention that. For new components / blocks, we now recommend that folks use function components, so I'd recommend going with that here as well.

simison commented:
FYI, it might be good to loop in Explorers team for review since they've done so much work in blocks (and related infra) here in Jetpack that they might have some framework level pointers.

Tug commented:
@jeherve @simison Thanks a lot for your initial review!
I've made some progress and I think it's starting to be in a good shape for a beta version. If you want to see it in action you can navigate to the [internal pre-alpha](https://wp.me/pbMoDN-GR) site where we're testing block creation from the mobile app: https://particular-wildcat.jurassic.ninja/.

I've tried to address most of your comments. Some I could not address relate to the technology choices.

First I would like to emphasise that I've designed the story player to be as independent from the jetpack codebase as possible. In the long term I don't see it live inside this repo. It will probably have it's own app served from wordpress.com (still to be decided as there's a possibility that Story becomes a core block as well).

About using Preact/HTM. I agree that it does increase the size of the block's `view.js`. I tried going vanillaJS at first but as the complexity grew, the code was getting harder to read and maintain and I was starting to add framework features manually to address that. In the end I feel preact/htm is a still good choice for this use case, it has a good balance of feature and size. I could have gone with any other light framework and the fact that you already use it at some other place in jetpack is merely a coincidence.

About SVGs I agree they're better for maintenance. In this case they're not our icons so we don't really need to maintain them. They're Materials icons, I just used the woff2 font version and filtered it to keep just the ones we needed. Inlining it into the CSS allows me to keep the build files count low (just one JS and one CSS) while not touching the webpack config. I don't think adding special cases for this app is worth it.

Thanks for suggesting me to loop in Explorer, I'm going to ping them today as I improve the PR description and list the things that are still left todo and might not make it for the beta.

simison commented:
> In the long term I don't see it live inside this repo. It will probably have it's own app served from wordpress.com (still to be decided as there's a possibility that Story becomes a core block as well).

Interesting! Note that if it's released to Jetpack users, you'll have to support it for a very long time since people will have it in their contents (it can be hidden from the block picker of course). I think Jetpack has a mechanism to ship blocks only to Simple and Atomic sites tho, Jeremy knows more.

Another possibility is to include it in FSE plugin (context p58i-8qD-p2) which is just for .com customers and also includes several blocks already.

simison commented:
>  I could have gone with any other light framework and the fact that you already use it at some other place in jetpack is merely a coincidence.

To this point, although `@wordpress/element` isn't lightweight per-se, when used across multiple features, the performance hit becomes shared and is less of an issue.

If you go with Preact, can we make sure that if Search bundles Preact and this bundles Preact, that it's loaded in the frontend just once?

> About SVGs I agree they're better for maintenance. In this case they're not our icons so we don't really need to maintain them.

By maintenance I also meant including new icons, removing icons or switching icons to something else.

Tug commented:
> If you go with Preact, can we make sure that if Search bundles Preact and this bundles Preact, that it's loaded in the frontend just once?


Oh when you say Search, it's on the frontend!? I didn't understand it that way but it makes sense now. I can look into it for sharing, it's probably doable if we tweak the webpack config. If we're using JSX on the frontend that might mean I could use it as well also, very interesting 👍

scruffian commented on extensions/blocks/story/player/utils.js:
```diff
@@ -0,0 +1,54 @@
+			document.msFullscreenEnabled
+		);
+	},
+	element() {
```
`enabled` and `element` seem to do the same thing?

scruffian replied:
Oh wait no they don't! I wonder if there's a way to refactor this file to be more DRY...

Tug commented:
I had a quick look at how search is built and I don't think I can easily adapt it for stories unless we make them into a jetpack module. It seems like quite some work for little gain imo: preact is 3KB and story is a media oriented block so I wouldn't worry too much about that extra payload at this point.
The fact that the app is shipped as part of the block view.js really makes our life easier at this point. I can understand how it's not the cleanest way to do things but can we see if this is really a blocker or not for this feature?

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,130 @@
+
+namespace Automattic\Jetpack\Extensions\Story;
+
+use Jetpack_AMP_Support;
```
You can remove this since we don't use it here.

```suggestion
```

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,130 @@
+function render( $attributes ) {
+	Jetpack_Gutenberg::load_assets_as_required( FEATURE_NAME );
+
+	$get_image_template = function( $media ) {
```
I'd recommend extracting those (image, video, slide templates) into their own function. It should make things easier to parse, and allow us to re-use each part later if needed.

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,130 @@
+
+	$get_image_template = function( $media ) {
+		if ( ! isset( $media['id'] ) || ! isset( $media['url'] ) ) {
+			return 'Error retrieving media';
```
I think it'd be nice if this were translatable. Same below.

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,130 @@
+	Jetpack_Gutenberg::load_assets_as_required( FEATURE_NAME );
+
+	$get_image_template = function( $media ) {
+		if ( ! isset( $media['id'] ) || ! isset( $media['url'] ) ) {
```
I think we could use `empty()` here, to cover more use-cases. Same below.

```suggestion
		if ( empty( $media['id'] ) || empty( $media['url'] ) ) {
```

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,130 @@
+			return 'Error retrieving media';
+		}
+		return sprintf(
+			'<img
+				alt="%s"
+				class="wp-block-jetpack-story_image wp-story-image wp-image-%d"
+				data-id="%d"
+				src="%s">',
+			esc_attr( $media['alt'] ),
+			$media['id'],
+			$media['id'],
+			esc_attr( $media['url'] )
```
I think you could move to numbered placeholders here, it would simplify things a bit:

```suggestion
			'<img
				alt="%1$s"
				class="wp-block-jetpack-story_image wp-story-image wp-image-%2$d"
				data-id="%2$d"
				src="%3$s">',
			esc_attr( $media['alt'] ),
			absint( $media['id'] ),
			esc_attr( $media['url'] )
```

Same below.

jeherve commented on extensions/blocks/story/player/components/header.js:
```diff
@@ -0,0 +1,35 @@
+		return null;
+	}
+
+	// default WP logo url relative to story/view.js
+	siteIconUrl = siteIconUrl || '../../../../../../wp-includes/images/w-logo-blue.png';
```
This is definitely going to be problematic, since the WordPress install does not always live there. We'd need to fetch the includes directory dynamically, or maybe pull an image that we ship with Jetpack instead? 

As a site owner, I think I'd be more interested in seeing my own site logo there, so maybe that's what we should try to use?

Tug replied:
You sure the directory structure could be different? This is a relative url so it should find the wp-includes dir relatively from the view.js script for the block. Since it's relative it should work even if wordpress is installed in a subdir. That said, user can always remove or rename the image so it's not super reliable.

Shipping this image with jetpack could make more sense I agree. I'm not sure what we'd want as a default, we need a square image and the logo isn't necessarily one afaik.

Tug replied:
Actually I'm already resolving this url in the PHP code [here](https://github.com/Automattic/jetpack/pull/16302/files#diff-b7f04975235c31f89dd45854dbc8079fR125) so I can ditch that code.

jeherve replied:
> You sure the directory structure could be different? 

Yeah, basically not all sites have plugins and themes live in `wp-content/`, so the directory structure of the site (and consequently the location of the block files compared to `wp-includes/`) is not always the same.

jeherve commented on extensions/blocks/story/player/components/header.js:
```diff
@@ -0,0 +1,35 @@
+			</div>
+			<SimpleButton
+				className="wp-story-exit-fullscreen"
+				label="Exit Fullscreen"
```
We'll need to translate this.

jeherve commented on extensions/blocks/story/index.js:
```diff
@@ -0,0 +1,75 @@
+	description: __( 'Add an interactive story.', 'jetpack' ),
+	attributes,
+	supports: {
+		html: false,
+		inserter: true, // toggle to false before merging to BETA blocks
```
Could we support alignment options too, since on desktop I think it'd be nice to have the slider take the whole content width for example.

Tug replied:
Yeah that's something we'll want to explore later. As of now it really is limited to a mobile-like format.

jeherve commented on extensions/blocks/story/index.js:
```diff
@@ -0,0 +1,75 @@
+ */
+import storyExample1 from './story_example-1.png';
+
+export const icon = (
+	<SVG xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24">
+		<Path d="M0 0h24v24H0z" fill="none" />
+		<Path
+			fillRule="evenodd"
+			clipRule="evenodd"
+			fill="#2C3338"
+			d="M6 3H14V17H6L6 3ZM4 3C4 1.89543 4.89543 1 6 1H14C15.1046 1 16 1.89543 16 3V17C16 18.1046 15.1046 19 14 19H6C4.89543 19 4 18.1046 4 17V3ZM18 5C19.1046 5 20 5.89543 20 7V21C20 22.1046 19.1046 23 18 23H10C8.89543 23 8 22.1046 8 21H18V5Z"
+		/>
+	</SVG>
+);
```
Since we're using a regular Play icon, do you think we could rely on one of the media dashicons instead? I think you could use the dashicon name when registering the block here in this file, and rely on `BlockIcon` with passing the dashicon name elsewhere.
https://developer.wordpress.org/resource/dashicons/

Tug replied:
I recently updated it, it's a custom icon now, are you sure you were testing the last version?

jeherve replied:
I think I was. How custom is it? It looks like a Play icon to me, but I'm no designer :)

Tug replied:
It should look like this 
![image](https://user-images.githubusercontent.com/230230/88517210-cc00ec00-cfee-11ea-8b0c-32e556f73216.png)

I replaced it a couple hours before your review though so you may have started the review without [this commit](https://github.com/Automattic/jetpack/pull/16302/commits/4dc7848cb412a02ea688776ad53deca67ea0b13b).

jeherve commented on extensions/blocks/story/player/components/bullet.js:
```diff
@@ -0,0 +1,30 @@
+		<button
+			key={ index }
+			className="wp-story-pagination-bullet"
+			aria-label={ `Go to slide ${ index }` }
```
We'll need to translate this as well.

Tug replied:
That's one advantage of going `@wordpress/element` instead of preact, I can load other gutenberg packages now such as `@wordpress/i18n` :). The downside is that the app is tied to the gutenberg/jetpack ecosystem now

jeherve commented on extensions/blocks/story/player/components/controls.js:
```diff
@@ -0,0 +1,26 @@
+	return (
+		<div className="wp-story-controls">
+			<SimpleButton
+				label="Play"
```
We'll need to translate those controls as well.

jeherve commented on extensions/blocks/story/editor.scss:
```diff
@@ -0,0 +1,55 @@
+		button {
+			width: 36px !important;
+			height: 36px !important;
+			color: #555d66;
+			border: 1px solid #555d66;
+			outline: 0;
+			background-color: transparent;
+
+			&:hover {
+				border: 1px solid #555d66;
+				background-color: transparent;
+				i {
+					color: rgb(51, 129, 184);
```
Can we rely on Core colors for these elements as well?

Tug replied:
Seems like this is the same styles defined by `components-button`. I still need to figure out how to have that class on the block editor and not on the frontend side

Tug replied:
Ok I would need to make use of the Button component in core and add some customized styles to make it round and so on. That would mean replacing the font with SVG (which is totally fine), removing my custom button components and updating the various instances. If we want to make it for the code freeze today I suggest I just add that to the tasks for next iteration if that's fine with you

jeherve replied:
That works for me 👍

jeherve commented on extensions/blocks/story/edit.js:
```diff
@@ -0,0 +1,138 @@
+		<Fragment>
+			{ controls }
+			{ noticeUI }
+			<div className="wp-block-jetpack-story wp-story">
```
I think we'd need to support custom CSS Classes here in the editor, as well as on the frontend. In the editor you have access to the `className` prop, and on the frontend, we have the `Jetpack_Gutenberg:block_classes()` utility that you can use.

Tug commented:
@jeherve Thanks for reviewing :)

> I think we would benefit from some kind of spinner or indicator that images are still loading when you upload images into the block.

Definitely, that's what I'm currently working on. The main issue with that is that currently all media are loaded by the browser on page load, instead I want to try to be smarter and load the first slides first then the second one. Regarding video I might just load them with an AJAX request instead since browsers usually do partial content request and never tries to load the whole video at once. Anyway that's a work in progress

> I don't think we need the Audio icon here?

You mean when we're playing images?

> Skipping to the next slide in the editor does not appear to work in the editor, whether you click or wait for the first slide to complete.

Seems like a bug, it might be related to the recent changes I did, I'll try to reproduce and fix before merging

> Upon loading the page on the frontend, the block immediately pops up full-screen. Should we wait until the reader clicks play to do that?

This happens when you load only 1 post that has a story https://github.com/Automattic/jetpack/pull/16302/files#diff-b7f04975235c31f89dd45854dbc8079fR90
Granted the UX could be revisited here. @megsfulton do you think we should remove this behavior?

Thanks for the comments on the code I'll address those 👍

megsfulton commented:
>Granted the UX could be revisited here. @megsfulton do you think we should remove this behavior?

Let's remove that behavior.

jeherve commented:
> You mean when we're playing images?

Yes. Since there is no sound we would not need an Audio icon there, would we?

Tug commented:
Good point! I think it's just meant to enable/disable sound for the whole story, not individual slides. We don't support sound over image at this point but that might be a feature, I'm not sure what's planned next ¯\_(ツ)_/¯

I'll forward your feedback to Megs and add it to [our board](https://github.com/orgs/Automattic/projects/191) though 👍

Tug commented:
Opened https://github.com/Automattic/jetpack/issues/16591 to track improvements on the story block.

jeherve commented:
> Opened #16591 to track improvements on the story block.

Now that we have an issue for follow-up items, what do you think about a rebase so we can merge this initial PR today, before the code freeze, and then iterations in follow-up PRs?

Tug commented:
@jeherve Awesome, I'll rebase now 👍

Tug commented:
> Skipping to the next slide in the editor does not appear to work in the editor, whether you click or wait for the first slide to complete.

I was able to reproduce this behavior only when media are still being uploaded. I will be addressing this as part of the "Show a progress indication on media upload" task.

Tug commented on package.json:
```diff
@@ -149,6 +149,7 @@
 		"react-pure-render": "1.0.2",
 		"react-redux": "6.0.1",
 		"react-router-dom": "5.2.0",
+		"react-shadow-dom-retarget-events": "1.0.11",
```
This is not needed anymore reverting...

jeryj commented:
Howdy! Here's a quick accessibility review. I'm sorry I didn't have more time to specifically address solutions for each. Here are the things I encountered when using the block with a keyboard and with VoiceOver in Safari.

### Frontend Story (not within modal)
- Play button shouldn't be a toggle, as you can't pause it from that button.
- You can tab to the long "Go to slide" buttons, but they don't do anything. I'd think either: 
    - pressing the long button opens the modal at that slide
    - or they are not buttons (or are disabled, but that can have its own tricky considerations at times, such as if it can switch from enabled to disabled, then focus needs to get managed so it's not lost to the top of the DOM)

### Frontend Story Overlay/Modal
- Story Overlay looks like a modal, but doesn't trap focus, so you can tab around and interact with the page behind the story modal. Turning this into a true modal should help address a few things, like the focus management, interactions (escape key) and an announcement that they are now in a modal when opening and leaving a modal when closing.
- Buttons don't have any focus indicator when they have focus - i.e. tabbing around,
- **Close button**: This doesn't need to be a toggle button.
- **Next/Previous buttons**: These also don't need to be toggle buttons. 
- **When going to the first/last slide**: The button that currenly had focus gets removed. Focus should be intentionally moved somewhere if the button gets removed or `disabled`. I'm not sure the best interaction here. There'll be a couple considerations around where to move focus to and how to make sure it doesn't interfere with announcement of the new content. Maybe moving focus _to_ the content with a `tabindex=-1`?
- **Play/Pause button**: It's a toggle and doesn't commmunicate its current state. This can be tricky. For one, the `aria-pressed` label doesn't update. This is honestly one that I'm not sure of the best solution. This article from [Heydon Pickering discusses play/pause toggles and considerations](https://inclusive-components.design/toggle-button/).
- **Progress slide buttons**
    - Instead of "Go to Slide", I think `Slide %d` should work here. IMO, the "Go to " is implied by it being a button. 
    - These are 0 indexed at the moment, so it announces "Go to slide 0" for the first image.
    - Nothing semantically indicates  which one is currently selected.
- **Moving to a new slide**: Nothing is announced. I'd expect announcing the new content, such as "Slide 2, [image alt text]"
- Escape key should close the modal and return focus from whence it came :)
- A lot of this looks like the carousel from the Newspack Carousel Block: https://github.com/Automattic/newspack-blocks/tree/master/src/blocks/carousel. You can use this on WordPress.com and see what the front-end implementation looks like. It's not perfect, but could help with general accessibility direction.

### Editor
- A lot of the same things from the Frontend will help here.
- It'd be nice to know that the slide was successfully added. 
- This may already be in the works, but why can't you use the Media Library to add a new slide? I would have assumed the same interface for adding new slides (Upload iage or select from media library)

In general, there's a lot of state changes and focus movements that happen, so it's good to keep in mind _how_:
- The current state is communicated
- When changing state, how to communicate the new state 

Let me know if you have any questions over any of this. This seems like a pretty neat feature 🎉

Tug commented:
Wow awesome! Thanks for the accessibility review @jeryj. I was going to look into it for next iteration but wasn't sure where to start. Your comments will help greatly :) May I bother you again once I have addressed all your points?
For this PR we'll want to merge so its available as a beta block on wpcom for internal testing. But this is on my high priority list for the improvements before we release it to the public https://github.com/Automattic/jetpack/issues/16591

jeryj commented:
@Tug Feel free to bother me anytime! :) I don't know how quickly I'll be able to get to another full-review once that time comes, but please always feel free to reach out with questions!

jeherve commented on extensions/blocks/story/player/use-resize-observer.js:
```diff
@@ -0,0 +1,102 @@
+	useRef,
+	useState,
+} from '@wordpress/element';
+import { useResizeObserver as useResizeObserverWordpress } from '@wordpress/compose';
```
Since `useResizeObserver` is not available in WordPress 5.4, you won't be able to rely on it without a check for it first, or you'll break things on sites that run WP 5.4.

Tug replied:
Yep the check happens on the next line actually

jeherve replied:
🤦 That will teach me to comment without paying attention. Thank you!

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,150 @@
+/**
+ * Story Block.
+ *
+ * @since 8.6.1
```
```suggestion
 * @since 8.8.0
```

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,150 @@
+	return sprintf(
+		'<img
+			alt="%1$s"
+			class="wp-block-jetpack-story_image wp-story-image wp-image-%2$s"
+			data-id="%2$s"
+			src="%3$s">',
+		esc_attr( $media['alt'] ),
+		$media['id'],
```
If the ID is an int, let's be explicit about it.

```suggestion
			class="wp-block-jetpack-story_image wp-story-image wp-image-%2$d"
			data-id="%2$d"
			src="%3$s">',
		esc_attr( $media['alt'] ),
		absint( $media['id'] ),
```

jeherve commented on extensions/blocks/story/story.php:
```diff
@@ -0,0 +1,150 @@
+		'<video
+			title="%1$s"
+			type="%2$s"
+			class="wp-story-video intrinsic-ignore wp-video-%3$s"
+			data-id="%3$s"
+			src="%4$s">
+		</video>',
+		esc_attr( $media['alt'] ),
+		esc_attr( $media['mime'] ),
+		$media['id'],
```
```suggestion
			class="wp-story-video intrinsic-ignore wp-video-%3$d"
			data-id="%3$d"
			src="%4$s">
		</video>',
		esc_attr( $media['alt'] ),
		esc_attr( $media['mime'] ),
		absint( $media['id'] ),
```

jeherve commented:
@Tug You should be good to commit the matching WordPress.com diff now so we can start testing things on WordPress.com sites as well.

jeherve commented:
r211125-wpcom

megsfulton commented:
>I think it's just meant to enable/disable sound for the whole story, not individual slides. We don't support sound over image at this point but that might be a feature, I'm not sure what's planned next ¯_(ツ)_/¯
>I'll forward your feedback to Megs and add it to our board though 👍

Good feedback! 

The audio icon controls the sound on a slide by slide basis. So for slides that are just images, the icon should not appear.
